core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -147,156 +146,105 @@ private static long optionalSum(long left, long right) {
+                size = queryResult.size();
+                TopDocs td = queryResult.topDocs();
+                if (td != null && td.scoreDocs.length > 0) {
+                    setShardIndex(td, queryResult.getShardIndex());
+                    topDocs.add(td);
+                }
+                Suggest shardSuggest = queryResult.suggest();
+                if (shardSuggest != null) {
+                    for (CompletionSuggestion suggestion : shardSuggest.filter(CompletionSuggestion.class)) {
+                        suggestion.setShardIndex(sortedResult.getShardIndex());
+                        List<Suggestion<CompletionSuggestion.Entry>> suggestions =
+                            groupedCompletionSuggestions.computeIfAbsent(suggestion.getName(), s -> new ArrayList<>());
+                        suggestions.add(suggestion);
                     }
-                    canOptimize = true;
-                    hasResult = true;
-                    resultToOptimize = entry.queryResult();
-                    shardIndex = resultToOptimize.getShardIndex();
                 }
             }
-            result = canOptimize ? resultToOptimize : results.stream().findFirst().get().queryResult();
-            assert result != null;
         }
-        if (canOptimize) {
-            int offset = result.from();
-            if (ignoreFrom) {
-                offset = 0;
-            }
-            ScoreDoc[] scoreDocs = result.topDocs().scoreDocs;
-            ScoreDoc[] docs;
-            int numSuggestDocs = 0;
-            final Suggest suggest = result.queryResult().suggest();
-            final List<CompletionSuggestion> completionSuggestions;
-            if (suggest != null) {
-                completionSuggestions = suggest.filter(CompletionSuggestion.class);
-                for (CompletionSuggestion suggestion : completionSuggestions) {
-                    numSuggestDocs += suggestion.getOptions().size();
-                }
-            } else {
-                completionSuggestions = Collections.emptyList();
-            }
-            int docsOffset = 0;
-            if (scoreDocs.length == 0 || scoreDocs.length < offset) {
-                docs = new ScoreDoc[numSuggestDocs];
-            } else {
-                int resultDocsSize = result.size();
-                if ((scoreDocs.length - offset) < resultDocsSize) {
-                    resultDocsSize = scoreDocs.length - offset;
-                }
-                docs = new ScoreDoc[resultDocsSize + numSuggestDocs];
-                for (int i = 0; i < resultDocsSize; i++) {
-                    ScoreDoc scoreDoc = scoreDocs[offset + i];
-                    scoreDoc.shardIndex = shardIndex;
-                    docs[i] = scoreDoc;
-                    docsOffset++;
+        if (size != -1) {
+            final ScoreDoc[] mergedScoreDocs = topDocs.isEmpty() ? EMPTY_DOCS :
+                mergeTopDocs(topDocs, size, ignoreFrom ? 0 : from).scoreDocs;
```

Comment:
Maybe we should still apply the single shard optim here by not calling mergeTopDocs if `topDocs.size() ==  1` ?